File: sync-server/src/utils/portable_path.rs
use crate::app_state::database::models::FileManifestEntries;
use anyhow::{Result, bail, ensure};
use std::collections::BTreeMap;
use unicode_normalization::UnicodeNormalization;

pub const INTERNAL_DIRECTORY: &str = ".vault-link-sync";

// Use the same locale-independent Unicode upper-case mapping in the client.
// NFC -> uppercase -> NFC also catches ß/SS and final/ordinary sigma aliases.
pub fn path_key(path: &str) -> String {
    path.nfc()
        .collect::<String>()
        .to_uppercase()
        .nfc()
        .collect()
}

pub fn validate_file_manifest(entries: &FileManifestEntries) -> Result<()> {
    // Store one component per node. Expanding every prefix makes a short,
    // deeply nested path consume quadratic memory and CPU before any I/O.
    let mut nodes: BTreeMap<(usize, String), (usize, &str, bool)> = BTreeMap::new();

    for path in entries.values() {
        ensure!(!path.is_empty(), "Path must not be empty");

        ensure!(
            path.nfc().collect::<String>() == *path,
            "Paths must be NFC normalized"
        );

        let parts: Vec<_> = path.split('/').collect();

        ensure!(
            path_key(parts[0]) != path_key(INTERNAL_DIRECTORY),
            "Reserved sync directory"
        );

        let mut parent = 0;
        for (index, part) in parts.iter().enumerate() {
            ensure!(
                !part.is_empty() && *part != "." && *part != "..",
                "Invalid path component"
            );

            ensure!(!part.ends_with(['.', ' ']), "Trailing dot or space in path");

            ensure!(
                !part
                    .chars()
                    .any(|c| c.is_control() || c == '\\' || "<>:\"|?*".contains(c)),
                "Non-portable filename"
            );

            let stem = part
                .split('.')
                .next()
                .unwrap_or_default()
                .trim_end_matches(' ')
                .to_uppercase();

            ensure!(
                !matches!(
                    stem.as_str(),
                    "CON" | "PRN" | "AUX" | "NUL" | "CONIN$" | "CONOUT$"
                ),
                "Reserved filename"
            );

            if stem.starts_with("COM") || stem.starts_with("LPT") {
                ensure!(
                    !matches!(
                        &stem[3..],
                        "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "¹" | "²" | "³"
                    ),
                    "Reserved filename"
                );
            }

            let is_file = index == parts.len() - 1;
            let key = (parent, path_key(part));

            if let Some((node, spelling, previous_is_file)) = nodes.get(&key) {
                if spelling != part || *previous_is_file || is_file {
                    bail!("Conflicting path: {path}");
                }
                parent = *node;
            } else {
                let node = nodes.len() + 1;
                nodes.insert(key, (node, part, is_file));
                parent = node;
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::{FileManifestEntries, path_key, validate_file_manifest};
    use uuid::Uuid;

    fn manifest(paths: &[&str]) -> FileManifestEntries {
        paths
            .iter()
            .enumerate()
            .map(|(index, path)| (Uuid::from_u128(index as u128), (*path).to_owned()))
            .collect()
    }

    #[test]
    fn path_keys_normalize_unicode_and_case_aliases() {
        for (path, expected) in [
            ("Notes/Café.md", "NOTES/CAFÉ.MD"),
            ("notes/Cafe\u{301}.md", "NOTES/CAFÉ.MD"),
            ("Straße.md", "STRASSE.MD"),
            ("σς.md", "ΣΣ.MD"),
        ] {
            assert_eq!(path_key(path), expected, "path: {path:?}");
        }
    }

    #[test]
    fn accepts_empty_manifests_and_portable_paths_with_shared_directories() {
        validate_file_manifest(&manifest(&[])).unwrap();

        let entries = manifest(&[
            "Notes/Café.md",
            "Notes/日記.md",
            "Notes/Archive/old.md",
            ".gitkeep",
            "COM10.txt",
            "LPT0.txt",
            ".vault-link-sync-backup/state.json",
            "Notes/.vault-link-sync/state.json",
        ]);

        validate_file_manifest(&entries).unwrap();
    }

    #[test]
    fn does_not_impose_a_cross_platform_total_path_limit() {
        for path in ["a".repeat(300), "é".repeat(200)] {
            validate_file_manifest(&manifest(&[&path])).unwrap();
        }
    }

    #[test]
    fn deep_paths_do_not_expand_every_prefix() {
        let path = vec!["a"; 10_000].join("/");
        let started = std::time::Instant::now();
        validate_file_manifest(&manifest(&[&path])).unwrap();
        assert!(
            started.elapsed() < std::time::Duration::from_secs(2),
            "A 20 KB path exhausted the validation work budget: {:?}",
            started.elapsed()
        );
    }

    #[test]
    fn rejects_invalid_components_characters_and_non_normalized_paths() {
        for path in [
            "",
            "/notes.md",
            "notes.md/",
            "notes//a.md",
            "./notes.md",
            "notes/../a.md",
            "notes./a.md",
            "notes /a.md",
            "Cafe\u{301}.md",
        ] {
            assert!(
                validate_file_manifest(&manifest(&[path])).is_err(),
                "accepted invalid path: {path:?}"
            );
        }

        for character in ['\\', '<', '>', ':', '"', '|', '?', '*', '\0', '\u{85}'] {
            let path = format!("notes/a{character}b.md");

            assert!(
                validate_file_manifest(&manifest(&[&path])).is_err(),
                "accepted non-portable character in path: {path:?}"
            );
        }
    }

    #[test]
    fn rejects_reserved_sync_and_windows_names() {
        for path in [
            ".vault-link-sync",
            ".VAULT-LINK-SYNC/state.json",
            "con",
            "prn.txt",
            "AUX.md",
            "nul .txt",
            "CONIN$",
            "conout$.log",
            "com1.txt",
            "LPT9",
            "COM¹.txt",
            "lpt²/file.md",
            "COM³.txt",
        ] {
            assert!(
                validate_file_manifest(&manifest(&[path])).is_err(),
                "accepted reserved path: {path:?}"
            );
        }
    }

    #[test]
    fn rejects_path_conflicts_in_either_entry_order() {
        for [first, second] in [
            ["notes.md", "notes.md"],
            ["Notes.md", "notes.md"],
            ["Straße.md", "STRASSE.md"],
            ["ς.md", "σ.md"],
            ["Notes/a.md", "notes/b.md"],
            ["notes", "notes/a.md"],
        ] {
            for paths in [[first, second], [second, first]] {
                assert!(
                    validate_file_manifest(&manifest(&paths)).is_err(),
                    "accepted conflicting paths: {paths:?}"
                );
            }
        }
    }
}

File: frontend/sync-client/src/utils/portable-path.ts
export const INTERNAL_DIRECTORY = ".vault-link-sync";
const MAX_COMPONENT_BYTES = 255;
const byteLength = (value: string): number =>
    new TextEncoder().encode(value).length;
function truncate(value: string, limit: number): string {
    let result = "";
    for (const character of value) {
        if (byteLength(result + character) > limit) break;
        result += character;
    }
    return result;
}

const reservedDeviceName =
    /^(CON|PRN|AUX|NUL|CONIN\$|CONOUT\$|COM[1-9¹²³]|LPT[1-9¹²³])$/u;

// NFC -> uppercase -> NFC matches the server's locale-independent alias rules.
const pathKey = (path: string): string =>
    path.normalize("NFC").toUpperCase().normalize("NFC");

export const arePathAliases = (left: string, right: string): boolean =>
    pathKey(left) === pathKey(right);

export function findPathWithSameSpelling(
    paths: Iterable<string>,
    wanted: string
): string | undefined {
    const normalized = wanted.normalize("NFC");
    for (const path of paths) {
        if (path !== wanted && path.normalize("NFC") === normalized)
            return path;
    }
    return undefined;
}

export const isInternalPath = (path: string): boolean =>
    arePathAliases(path.split("/")[0], INTERNAL_DIRECTORY);

function validatePath(path: string): void {
    if (!path || path !== path.normalize("NFC") || isInternalPath(path))
        throw new Error(`Invalid portable path: ${path}`);

    for (const part of path.split("/")) {
        if (
            !part ||
            part === "." ||
            part === ".." ||
            /[<>:"\\|?*\p{Cc}]|[. ]$/u.test(part) ||
            reservedDeviceName.test(
                part.split(".")[0].replace(/ +$/u, "").toUpperCase()
            )
        )
            throw new Error(`Invalid portable path: ${path}`);
    }
}

export function validatePortablePaths(paths: Iterable<string>): void {
    // Component keys keep storage proportional to the input, even when a
    // manifest contains thousands of directory levels.
    const nodes = new Map<
        string,
        { id: number; spelling: string; file: boolean }
    >();
    for (const path of paths) {
        validatePath(path);
        const parts = path.split("/");
        let parent = 0;
        parts.forEach((spelling, index) => {
            const file = index === parts.length - 1;
            const key = `${parent}/${pathKey(spelling)}`;
            const previous = nodes.get(key);
            if (
                previous &&
                (previous.spelling !== spelling || previous.file || file)
            )
                throw new Error(`Conflicting path: ${path}`);
            const node = previous ?? { id: nodes.size + 1, spelling, file };
            nodes.set(key, node);
            parent = node.id;
        });
    }
}

function portableName(part: string): string {
    let result = part
        .normalize("NFC")
        .replace(/[<>:"\\|?*\p{Cc}]/gu, "_")
        .replace(/[. ]+$/u, "_");
    if (!result || result === "." || result === "..") result = "_";
    if (
        reservedDeviceName.test(
            result.split(".")[0].replace(/ +$/u, "").toUpperCase()
        )
    )
        result = "_" + result;
    return result;
}

function fitName(name: string, suffix: string, file: boolean): string {
    const dot = file ? name.lastIndexOf(".") : -1;
    const extension = dot > 0 ? truncate(name.slice(dot), 64) : "";
    const stem = dot > 0 ? name.slice(0, dot) : name;
    return portableName(
        truncate(stem, MAX_COMPONENT_BYTES - byteLength(suffix + extension)) +
            suffix +
            extension
    );
}

function pathsConflict(left: string, right: string): boolean {
    const a = left.split("/"),
        b = right.split("/");
    for (let i = 0; i < Math.min(a.length, b.length); i++) {
        if (!arePathAliases(a[i], b[i])) return false;
        if (a[i] !== b[i]) return true;
    }
    return true; // Equal names, or a file occupies an ancestor directory.
}

function isAvailablePath(
    candidate: string,
    id: string,
    occupied: Readonly<Record<string, string>>
): boolean {
    validatePath(candidate);
    return Object.entries(occupied).every(
        ([otherId, path]) => otherId === id || !pathsConflict(candidate, path)
    );
}

/** Deterministic allocation also handles a file blocking an ancestor directory. */
export function allocatePortablePath(
    wanted: string,
    id: string,
    occupied: Readonly<Record<string, string>>
): string {
    id = portableName(truncate(id, 64));
    const originals = wanted.split("/");
    const parts = originals.map((part, index) => {
        const portable = portableName(part);
        return byteLength(portable) > MAX_COMPONENT_BYTES
            ? fitName(
                  portable,
                  ` (conflict ${id})`,
                  index === originals.length - 1
              )
            : portable;
    });
    if (isInternalPath(parts.join("/"))) parts[0] = "_" + parts[0];
    let candidate = parts.join("/");
    const suffix = ` (conflict ${id})`;
    const limit = (Object.keys(occupied).length + 1) * (parts.length + 1);
    for (let attempt = 0; attempt <= limit; attempt++) {
        try {
            if (isAvailablePath(candidate, id, occupied)) return candidate;
        } catch {
            /* allocate below */
        }
        const current = candidate.split("/");
        let conflictAt = current.length - 1;
        for (let i = 0; i < current.length; i++) {
            const prefix = current.slice(0, i + 1).join("/");
            if (
                Object.values(occupied).some((path) => {
                    const other = path
                        .split("/")
                        .slice(0, i + 1)
                        .join("/");
                    return (
                        arePathAliases(prefix, other) &&
                        (prefix !== other ||
                            path.split("/").length === i + 1 ||
                            i === current.length - 1)
                    );
                })
            ) {
                conflictAt = i;
                break;
            }
        }
        const original = parts[conflictAt] ?? current[conflictAt];
        current[conflictAt] = fitName(
            original,
            suffix + (attempt ? ` (${attempt})` : ""),
            conflictAt === current.length - 1
        );
        candidate = current.join("/");
    }
    throw new Error(`Unable to allocate a portable path: ${wanted}`);
}
